components/engine_rocks/src/write_batch.rs

```diff
@@ -132,6 +152,174 @@ impl Mutable for RocksWriteBatch {
     }
 }
 
+/// `RocksWriteBatchVec` is for method `MultiBatchWrite` of RocksDB, which splits a large WriteBatch
+/// into many smaller ones and then any thread could help to deal with these small WriteBatch when it
+/// is calling `MultiBatchCommit` and wait the front writer to finish writing. `MultiBatchWrite` will
+/// perform much better than traditional `pipelined_write` when TiKV writes very large data into RocksDB.
+/// We will remove this feature when `unordered_write` of RocksDB becomes more stable and becomes compatible
+/// with Titan.
+pub struct RocksWriteBatchVec {
+    db: Arc<DB>,
+    wbs: Vec<RawWriteBatch>,
+    save_points: Vec<usize>,
+    index: usize,
+    cur_batch_size: usize,
+    batch_size_limit: usize,
+}
+
+impl RocksWriteBatchVec {
+    pub fn new(db: Arc<DB>, batch_size_limit: usize, cap: usize) -> RocksWriteBatchVec {
+        let wb = RawWriteBatch::with_capacity(cap);
+        RocksWriteBatchVec {
+            db,
+            wbs: vec![wb],
+            save_points: vec![],
+            index: 0,
+            cur_batch_size: 0,
+            batch_size_limit,
+        }
+    }
+
+    pub fn as_inner(&self) -> &[RawWriteBatch] {
+        &self.wbs[0..=self.index]
+    }
+
+    pub fn as_raw(&self) -> &RawWriteBatch {
+        &self.wbs[0]
+    }
+
+    pub fn get_db(&self) -> &DB {
+        self.db.as_ref()
+    }
+
+    /// `check_switch_batch` will split a large WriteBatch into many smaller ones. This is to avoid
+    /// a large WriteBatch blocking write_thread too long.
+    fn check_switch_batch(&mut self) {
+        if self.batch_size_limit > 0 && self.cur_batch_size >= self.batch_size_limit {
+            self.index += 1;
+            self.cur_batch_size = 0;
+            if self.index >= self.wbs.len() {
+                self.wbs.push(RawWriteBatch::default());
+            }
+        }
+        self.cur_batch_size += 1;
```

Comment:
why add 1 here since set it to 0 in above if block?

Reply:
As whether we need to switch or not, we always update `cur_batch_size` before performing `put`, `delete`, etc. operations.